experimental/unicodeset_parser/src/parse.rs

```diff
@@ -169,6 +184,47 @@ impl ParseError {
     }
 }
 
+/// The value of a variable in a UnicodeSet. Used as value type in [`VariableMap`].
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub enum VariableValue<'a> {
+    /// A UnicodeSet, represented as a [`CodePointInversionListAndStringList`](CodePointInversionListAndStringList).
+    UnicodeSet(CodePointInversionListAndStringList<'a>),
+    // in theory, a one-code-point string is always the same as a char, but we might want to keep
+    // this variant for efficiency?
+    /// A single code point.
+    Char(char),
+    /// A string.
+    String(String),
+}
+
+impl From<char> for VariableValue<'_> {
+    fn from(c: char) -> Self {
+        VariableValue::Char(c)
+    }
+}
+
+impl From<String> for VariableValue<'_> {
+    fn from(s: String) -> Self {
+        VariableValue::String(s)
+    }
+}
+
+impl From<&str> for VariableValue<'_> {
+    fn from(s: &str) -> Self {
+        VariableValue::String(s.to_owned())
+    }
+}
+
+impl<'a> From<CodePointInversionListAndStringList<'a>> for VariableValue<'a> {
+    fn from(set: CodePointInversionListAndStringList<'a>) -> Self {
+        VariableValue::UnicodeSet(set)
+    }
+}
+
+/// The map used for parsing UnicodeSets with variable support. See [`parse_with_variables`].
+pub type VariableMap<'a> = HashMap<String, VariableValue<'a>>;
```

Comment:
Use a newtype here. Type alias on public APIs can be confusing and commits us to a specific representation. You probably only need to redirect `insert` and `remove`.

Reply:
Makes sense, thanks!